pyhf/pdf.py

```diff
@@ -131,6 +137,173 @@ def __init__(self, spec, **config_kwargs):
+        _allchannels = list(set(_allchannels))
+        self.do_samples  = _allsamples[:]
+        self.do_channels = _allchannels[:]
+        self.do_mods = _allmods[:]
+        self.channel_nbins = channel_nbins
+
+        self.finalized_stats = {k:finalize_stats(self.config.modifier(k)) for k,v in self.config.par_map.items() if 'staterror' in k}
+        self._make_mega()
+        self.prep()
+
+    def _make_mega(self):
+        helper = {}
+        for c in self.spec['channels']:
+            for s in c['samples']:
+                helper.setdefault(c['name'],{})[s['name']] = (c,s)
+
+        mega_mods = {}
+        import copy
+        for m,mtype in self.do_mods:
+            for s in self.do_samples:
+                modspec = {'type': mtype, 'name': m}
+                if mtype == 'histosys':
+                    modspec.setdefault('data',{})['hi_data'] = []
+                    modspec.setdefault('data',{})['lo_data'] = []
+                    modspec.setdefault('data',{})['mask'] = []
+                elif mtype == 'normsys':
+                    modspec.setdefault('data',{})['hi'] = []
+                    modspec.setdefault('data',{})['lo'] = []
+                    modspec.setdefault('data',{})['mask'] = []
+                elif mtype == 'normfactor':
+                    modspec.setdefault('data',{})['mask'] = []
+                elif mtype == 'shapefactor':
+                    modspec.setdefault('data',{})['mask'] = []
+                elif mtype == 'shapesys':
+                    modspec.setdefault('data',{})['mask'] = []
+                elif mtype == 'staterror':
+                    modspec.setdefault('data',{})['uncrt'] = []
+                    modspec.setdefault('data',{})['mask']  = []
+                mega_mods.setdefault(s,{})[m] = copy.deepcopy(modspec)
+                
+        mega_samples = {}
+        for s in self.do_samples:
+            mega_nom = []
+            for c in self.do_channels:
+                defined_samp = helper.get(c,{}).get(s)
+                defined_samp = None if not defined_samp else defined_samp[1]
+                nom = defined_samp['data'] if defined_samp else [0.0]*self.channel_nbins[c]
+                mega_nom += nom
+                defined_mods = {x['name']:x for x in defined_samp['modifiers']} if defined_samp else {}
+                for m,mtype in self.do_mods:
+                    thismod = defined_mods.get(m)
+                    if mtype == 'histosys':
+                        lo_data = thismod['data']['lo_data'] if thismod else nom
+                        hi_data = thismod['data']['hi_data'] if thismod else nom
+                        maskval = True if thismod else False
+                        mega_mods[s][m]['data']['lo_data'] += lo_data
+                        mega_mods[s][m]['data']['hi_data'] += hi_data
+                        mega_mods[s][m]['data']['mask']    += [maskval]*len(nom) #broadcasting
+                        pass
```

Comment:
for histogram systematics, samples that are affected, store `hi_data` `lo_data` we add this to the corresponding `hi_data` and `lo_data` of the mega_histosys of that name. if the sample is not originally affected, we add dummy data where hi_data=lo_data=nom such that the delta of the interpolation will always be zero (we also keep a mask, so that we can actually force it to zero to in case we run into numerical issues